transformer_engine/common/permutation/permutation.cu

```diff
@@ -0,0 +1,330 @@
+      } else if (topK <= 64) {
+        moe_permute_kernel<T, TCompute, 64, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, topK, num_cols);
+      } else if (topK <= 128) {
+        moe_permute_kernel<T, TCompute, 128, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, topK, num_cols);
+      } else {
+        NVTE_ERROR("topK cannot exceed 128.");
+      }
+    }
+  }
+}
+
+template <typename T>
+void nvte_unpermute_launcher(const T *input, T *output, int *row_id_map, const float *prob,
+                             const int num_rows, const int topK, const int num_cols,
+                             cudaStream_t stream) {
+  using TCompute = typename std::conditional<(std::is_same<T, __nv_fp8_e5m2>::value ||
+                                              std::is_same<T, __nv_fp8_e4m3>::value),
+                                             half, T>::type;
+
+  static constexpr int kElementsPerAccess = 16 / sizeof(T);
+
+  int blocks = num_rows;
+  int threads = std::min(num_cols / kElementsPerAccess, 1024);
+  size_t smem_bytes = topK * sizeof(TCompute);
+
+  if (prob == nullptr) {
+    // moe_permute_bwd
+    // moe_unpermute_fwd without probs
+
+    moe_unpermute_kernel<T, TCompute, false><<<blocks, threads, smem_bytes, stream>>>(
+        input, output, row_id_map, nullptr, num_rows, topK, num_cols);
+  } else {
+    // moe_unpermute_fwd with probs
+
+    moe_unpermute_kernel<T, TCompute, true><<<blocks, threads, smem_bytes, stream>>>(
+        input, output, row_id_map, prob, num_rows, topK, num_cols);
+  }
+}
+
+void nvte_permute(const void *input, void *output, const transformer_engine::DType dtype,
+                  const int *sorted_row_id, int *row_id_map, const float *prob, const int num_rows,
+                  const int topK, const int num_cols, const int num_out_tokens, float *prob_grad,
+                  const void *input_fwd, cudaStream_t stream) {
+  TRANSFORMER_ENGINE_TYPE_SWITCH_ALL(
+      dtype, T,
+      nvte_permute_launcher(reinterpret_cast<const T *>(input), reinterpret_cast<T *>(output),
+                            sorted_row_id, row_id_map, prob, num_rows, topK, num_cols,
+                            num_out_tokens, prob_grad, reinterpret_cast<const T *>(input_fwd),
+                            stream););
+}
+
+void nvte_unpermute(const void *input, void *output, const transformer_engine::DType dtype,
+                    int *row_id_map, const float *prob, const int num_rows, const int topK,
+                    const int num_cols, cudaStream_t stream) {
+  TRANSFORMER_ENGINE_TYPE_SWITCH_ALL(
+      dtype, T,
+      nvte_unpermute_launcher(reinterpret_cast<const T *>(input), reinterpret_cast<T *>(output),
+                              row_id_map, prob, num_rows, topK, num_cols, stream););
+}
```

Comment:
TE does provide C/C++ compatible APIs and these `nvte_permute` and `nvte_unpermute` should be among those. 

Please follow the TE code design, these APIs should take `NVTETensor` as inputs instead of buffer pointers. This is especially important for FP8 and future datatypes as the `NVTETensor` carries more information than just a data ptr. 

Please have a look here https://github.com/NVIDIA/TransformerEngine/blob/0b303dad4c66693351dda93e11c7b61b873751b0/transformer_engine/common/transpose/transpose.cu#L296-L302

Reply:
Done.